### DISCUSSION TITLE
Drawing a fullscreen quad without buffers

### DISCUSSION THREAD

**Discussion by junglie85** [gfx-rs/wgpu#3276] [Q&A] [answered]
I'm trying to implement [drawing a fullscreen quad without buffers](https://www.saschawillems.de/blog/2016/08/13/vulkan-tutorial-on-rendering-a-fullscreen-quad-without-buffers/), as described by Sascha Willems.

However, I can't figure out how to do the left shift, which currently just errors:

```
struct VsIn {
    @builtin(vertex_index) vertex_index: u32,
};

struct VsOut {
    @builtin(position) position: vec4<f32>,
    @location(0) tex_coords: vec2<f32>,
};

@vertex
fn vs_main(in: VsIn) -> VsOut {
    let vertex_index = i32(in.vertex_index);
    let tex_coords = vec2<f32>((vertex_index << 1) & 2, vertex_index & 2);
    let position = vec4<f32>(tex_coords * 2.0 - 1.0, 0.0, 1.0);

    return VsOut(position, tex_coords);
}

---

thread 'main' panicked at 'wgpu error: Validation Error

Caused by:
    In Device::create_shader_module
      note: label = `color shader`

Shader validation error:
   ┌─ color shader:28:18
   │
28 │     let x = f32((vertex_index << 1) & 2);
   │                  ^^^^^^^^^^^^^^^^^ naga::Expression [5]


    Entry point vs_main at Vertex is invalid
    Expression [5] is invalid
    Operation ShiftLeft can't work with [3] and [4]
```

I don't find the explanation of shift left in the WGSL spec to be very helpful in explaining how to use this operator or it's valid types. How can I implement this in WGSL?

**Comment by cwfitzgerald** [comment, marked as answer]
The answer to your woes is unfortunately extremely boring - suffix all your numbers with `u` to make them unsigned. Currently Naga hasn't adopted the newer wgsl spec that allows for integer literal type deduction.

Errors are something we're aware of, and slowly making progress on.

**Comment by junglie85** [comment]
Perfect, thank you.

In case anyone else is looking for the solution, I went with:

```
@vertex
fn vs_main(in: VsIn) -> VsOut {
    let u = f32((in.vertex_index << 1u) & 2u);
    let v = f32(in.vertex_index & 2u);
    let tex_coords = vec2<f32>(u, v);
    
    let position = vec4<f32>(tex_coords * 2.0 - 1.0, 0.0, 1.0);
    let tex_coords = vec2<f32>(tex_coords.x, 1.0 - tex_coords.y); // Invert y-axis due to WGPU's y-axis being up not down.

    return VsOut(position, tex_coords);
}
```
